### DISCUSSION TITLE
Qt6 plans?

### DISCUSSION THREAD

**Discussion by peabee** [lxqt/lxqt#1901] [Q&A] [answered]
Qt6 is released - https://www.qt.io/blog/qt-6.0-released on 8 Dec 2020 ...........

Are there any plans to move LXQt to Qt6? Will such a move coincide with LXQt-0.17.0?

Thanks.

**Comment by tsujan** [comment]
Whenever Qt6 comes to our distros, we'll use it to develop LXQt ~~and so, LXQt will be compatible with it in its next release — or sooner, in its git version~~.  LXQt will be compatible with it in its next release only if Qt6 is ready for use (Qt6.0.0 isn't).

If Qt6 is backward compatible with Qt 5.15, the latter will be supported too. If not, Qt5 support might be removed from the next release (April 2021).

**Comment by peabee** [comment]
I think Qt6 has arrived already on ArchLinux....
https://archlinux.org/groups/x86_64/qt6/
and ArchLinux32....
https://www.archlinux32.org/packages/?sort=&q=qt6

**Reply by tsujan** [reply]
Yes, it'll come soon to other rolling distros too,

Having read some pages about Qt6 changes, I think the transition will be straightforward and boring —  boring because we should remove all deprecated stuff from our codes.

EDIT: I was wrong. The transition won't be easy; many things have changed in Qt6.

**Comment by luis-pereira** [comment]
Started the port of lxqt-build-tools. It's straightforward. Currently missing only the translations modules. 
The port will be easy and boring.

**Reply by tsujan** [reply]
Thanks! Will read your patches as soon as I find the time.

**Reply by tsujan** [reply]
I don't have Qt6 yet but it seems that x11extras is removed from it. As far as I remember, we use `QX11Info` in several places, especially in lxqt-panel.

These questions arise: Why did we use `QX11Info` when we had `KWindowSystem`? Is there anything that the former provides while the latter doesn't? If so, how could we replace them? Using X11 directly isn't a good idea.

I just hope that those use cases could be replaced by what `KWindowSystem` provides; otherwise, we'll have serious troubles.

EDIT:

After searching the Internet for what is done in KDE, I think we don't need to hurry. KF6 doesn't seem to be ready soon.

All we need to do is update our codes to Qt 5.15 and wait for KF6.

**Reply by luis-pereira** [reply]
Most of the code using `QX11Info` was written before `KWindowSystem` was used. It never got ported to KWindowSystem.
Happy New Year.

**Reply by tsujan** [reply]
Happy new year :)

> Most of the code using QX11Info was written before KWindowSystem was used.

There are some `QX11Info` methods that `KWindowSystem` doesn't provide, maybe, exactly because they're provided by `QX11Info`.

We'll see what Qt and/or KDE devs will do. Although Qt6 has come, a Qt6 port is impossible for now and I think it can't happen soon.

**Reply by yan12125** [reply]
As a record, Qt 6.2.0 beta 2 includes `qtx11extras_p.h` [1], and it has a `QX11Info` class [2], which seems similar to the one from Qt 5.

[1] https://archlinux.org/packages/kde-unstable/x86_64/qt6-base/files/
[2] https://github.com/qt/qtbase/blob/v6.2.0-beta2/src/gui/platform/unix/qtx11extras_p.h

**Reply by tsujan** [reply]
> As a record, Qt 6.2.0 beta 2... has a QX11Info class 

Thanks for the info! I'll make some of my apps work with Qt6 when 6.2.0 gets stable.

We can't do it before the next LXQt release though (and I think it's better not to do it either).

**Comment by Chiitoo** [comment]
For the 'extras', it seems the functionality should be available via other modules, though I can't point out exactly where.

See also: https://www.qt.io/blog/add-on-support-in-qt-6.0-and-beyond

**Reply by tsujan** [reply]
Thanks!

I don't like it; have a bad feeling about Qt's future... Anyway, we have to wait until 6.2, at least.

**Reply by DAP-DarkneSS** [reply]
@tsujan , Qt 6.2 is ready ;)

**Reply by tsujan** [reply]
Maybe with the LXQt version after the next. No distro is in a rush; we aren't either.

**Comment by tsujan** [comment]
OK, after wrestling with Qt6.0.0, I can say with certainty: **Qt6.0.0 is far from being ready for use.**

My first-hand experience, while porting FeatherPad to Qt6:

 * First of all, I had to sacrifice several functionalities because Qt6 doesn't support them (yet?).
 * Several classes are moved to `core5compat`. That wasn't an issue in itself. The problem was that some important methods were totally removed. Therefore, I had to sacrifice more functionalities.
 * There were lots of warnings that didn't exist in Qt5.15. I succeeded in silencing them but the fact is that removing all warnings with Qt5.15 doesn't guarantee anything.
 * Some methods were changed but the compiler considered them as fatal errors. I found the new methods after searching in Qt6 doc.
 * At last, FeatherPad was compiled cleanly against Qt6. I ran it and its window was opened correctly. At first, it seemed that its remaining functionalities were OK but that was a false hope. Weird things could be seen here and there. In short, it wasn't usable.

I don't remember if similar things happened with Qt5.0.0 because, in that time, I used Debian Testing and started porting my apps to Qt5 when it was mature enough.

EDIT: Strangely enough, porting of Kvantum to Qt6 needed only one sacrifice, namely the blur effect, which isn't important and could be restored when `KF6WindowSystem` is released. Perhaps Qt devs still haven't had time to touch `QStyle`...

**Reply by Borwe** [reply]
I am curious, what base distro are you using currently? Was wondering because I wanna see options to get qt5.15 atleast, but not on arch or testing debian, and yet be able to build and use lxqt on it.

**Reply by tsujan** [reply]
> ... but not on arch

I may have not understood your question but Qt 6.0.0 came to Arch in December 2020. The really used Qt is still Qt 5.15.

**Reply by darkshram** [reply]
Qt6 6.2.0 was released today (September 30, 2021). As suspected, it did not include qtx11extras. [README.md](https://github.com/qt/qtx11extras/blob/dev/README.md) at qtx11extras github says module it's deprecated and functionality has been incorporated into other parts of Qt, or in some cases removed because it was obsolete (something similar happened to GTK4, they deprecated all x11 specific widgets and will be removed in later releases). So. basically they say projects should move away from qtx11extras and [use alternative APIs](https://doc.qt.io/qt-6/portingguide.html).

New modules ported to Qt6: qtmultimedia, qtconnectivity, qtsensors, qtremoteobjects, qtserialport, qtserialbus, qtwebchannel, qtwebsockets and the long awaited qtwebengine.

Other interesting changes:
- qtquickcontrols2 was incorporated to qtdeclarative, which now requires qtshadertools
- Qt6 6.2.0 builds with GCC >= 8.3 (fails to build with GCC 7.3.1 because incomplete c++17 features) or a relatively recent version of Clang (supports full c++17 since version 5.0)

So, guess a stable KF6 release is not far away in the calendar.

**Reply by tsujan** [reply]
> As suspected, it did not include qtx11extras

KDE devs definitely need the same methods of qtx11extras that we do (just a few). We'll see what they do.

**Reply by tsujan** [reply]
For the record:

It seems the answer is [here](https://doc-snapshots.qt.io/qt6-dev/qguiapplication.html#nativeInterface) (the page address might change later):
```c++
QNativeInterface::QX11Application
```
Then, the X connection and display can be known from https://doc-snapshots.qt.io/qt6-dev/qnativeinterface-qx11application.html. I think that's all we need.

EDIT:

Sample code:
```c++
if (auto x11NativeInterfce = qApp->nativeInterface<QNativeInterface::QX11Application>()) {
    Display *disp = x11NativeInterfce->display();
}
```

**Comment by cats-works** [comment]
qt 6.2 is far far more in line with qt5's compatibility than 6.0/6.1 was.

i am working on qtermwidget's port to qt6.  it will initially require core5compat just for qtextcodec support (qregexp is ported to qregularexpression).  qt6 has very very little support for codecs aside from utf8/utf16/latin1 and i find this a bad regression.  doesn't look like the KDE team has much of an answer yet either on the subject.

it compiles and builds and "runs" but there is some weird behaviors currently (like backspace/delete not functioning) on the same code where it works fine on qt5.  i'll send a pull request when it's all good:  https://github.com/kitsnotes/qtermwidget-qt6

**Reply by yan12125** [reply]
Oops, I have started a working-in-progress branch https://github.com/lxqt/qtermwidget/tree/wip-qt6 but apparently I didn't indicate my progress clearly. Sorry for some duplicataed efforts.

There are more discussions about qtermwidget and other components in https://github.com/lxqt/lxqt/discussions/2139, by the way.

**Reply by cats-works** [reply]
thanks @yan12125 and don't be sorry that's on me for not really looking at the branches more carefully :)  it does look like we have been on the same path so let me finish up a few dangling deprecated function warnings and then i'll submit the PR to at least get things building.

i'll admit that this is selfishly motivated - i am working on another project where qt6 is standard from the get-go.  i'm not super keen on forking qtermwidget for my terminal.  i'd much rather have cooperation and use upstream where all possible :)

thanks,
~cat

**Comment by redtide** [comment]
Any news about this? I rarely use qt6 (I'm on Arch, still using 5) but I see that after a build the window have no theme, nor dark mode (I'm using a dark theme with Kvantum). I guess the reason is because this is still in progress?
The app I'm currently working on is a semi-rewrite of obconf-qt (to work for both Openbox under X11 and labwc on Wayland), so I had to work around that QX11Info issue too.

**Reply by tsujan** [reply, marked as answer]
Qt6 has been supported for a long time. Compile and install the Qt6 branches (→ https://github.com/lxqt/lxqt/discussions/2139)
